### PR TITLE
Add RSpec/ExpectOutput cop

### LINKED ISSUE #277: Cop Idea: Prefer `expect { ... }.to output` over `before { $stdout = StringIO.new }`
Instead of doing

```ruby
before do
  $stdout = StringIO.new
  $stderr = StringIO.new
end

after do
  $stdout = STDOUT
  $stderr = STDERR
end

it 'writes some output' do
  execute_code

  expect($stdout.read).to eql(...)
end
```

You should do

```ruby
it 'writes some output' do
  expect { execute_code }.to output(...).to_stdout
end
```

### PR DESCRIPTION
**Pull request by backus**
Fixes #277

### CODE REVIEW

File: lib/rubocop/cop/rspec/expect_output.rb [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+      #
+      # @example
+      #   # bad
+      #   $stdout = StringIO.new
+      #   my_app.print_report
+      #   $stdout = STDOUT
+      #   expect($stdout.string).to eq('Hello World')
+      #
+      #   # good
+      #   $stdout = StringIO.new
```
Comment (bquorning) [inline lib/rubocop/cop/rspec/expect_output.rb]: This line is "good"?

Comment (backus) [inline lib/rubocop/cop/rspec/expect_output.rb]: whoops

**Comment by backus** [comment]
@bquorning what do you think of this? I'm interested in looking into possible autocorrect but that can wait for a followup PR.

File: lib/rubocop/cop/rspec/expect_output.rb (hunk 1 above)
Comment (bquorning) [inline lib/rubocop/cop/rspec/expect_output.rb]: :-)

File: spec/rubocop/rspec/hook_spec.rb [hunk 2]
```diff
@@ -32,6 +32,10 @@ def hook(source)
     expect(hook('before(:each) { example_setup }').valid_scope?).to be(true)
   end
 
+  it 'classies :each as an example hook' do
```
Comment (bquorning) [inline spec/rubocop/rspec/hook_spec.rb]: typo: classies -> classifies

File: spec/rubocop/cop/rspec/expect_output_spec.rb [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+
+  it 'does not flag assignments to $stdout outside of example scope' do
+    expect_no_violations(<<-RUBY)
+      before(:suite) do
+        $stderr = StringIO.new
+      end
+    RUBY
+  end
+
+  it 'does not flag assignments to $stdout outside in example_group scope' do
```
Comment (bquorning) [inline spec/rubocop/cop/rspec/expect_output_spec.rb]: typo: remove "outside"

File: lib/rubocop/cop/rspec/expect_output.rb [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+      #   $stdout = STDOUT
+      #   expect($stdout.string).to eq('Hello World')
+      #
+      #   # good
+      #   expect { my_app.print_report }.to output('Hello World').to_stdout
+      class ExpectOutput < Cop
+        MSG = 'Use `expect { ... }.to output(...).to_%<name>s` ' \
+              'instead of mutating $%<name>s'.freeze
+
+        def_node_matcher :example?, Examples::ALL.block_pattern
```
Comment (bquorning) [inline lib/rubocop/cop/rspec/expect_output.rb]: I think `:example?` can be extracted to `Language::NodePattern` by now.

File: lib/rubocop/cop/rspec/expect_output.rb [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+        def_node_matcher :hook?, Hooks::ALL.block_pattern
+
+        def on_gvasgn(node)
+          return unless inside_example_scope?(node)
+
+          variable_name, = *node
```
Comment (bquorning) [inline lib/rubocop/cop/rspec/expect_output.rb]: I really like when we assign to underscore-prefixed variables. When writing new cops, it is nice to be able to look at the existing ones and see what kind of arguments are passed into the various `on_` methods.

File: lib/rubocop/cop/rspec/expect_output.rb [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+
+          add_offense(node, :name, format(MSG, name: name))
+        end
+
+        private
+
+        def inside_example_scope?(node)
+          return false if node.nil? || example_group?(node)
+          return true if example?(node)
+          return RuboCop::RSpec::Hook.new(node).example? if hook?(node)
```
Comment (bquorning) [inline lib/rubocop/cop/rspec/expect_output.rb]: This line I don’t understand.

File: spec/rubocop/cop/rspec/expect_output_spec.rb [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+  it 'registers an offense for overwriting $stdout' do
+    expect_violation(<<-RUBY)
+      specify do
+        $stdout = StringIO.new
+        ^^^^^^^ Use `expect { ... }.to output(...).to_stdout` instead of mutating $stdout
+      end
+    RUBY
+  end
+
+  it 'registers an offense for overwriting $stderr' do
```
Comment (bquorning) [inline spec/rubocop/cop/rspec/expect_output_spec.rb]: The description doesn’t properly describe the difference from the previous spec example. What’s special about `before(:each)` compared to `before(:suite)` in line 44?

File: lib/rubocop/cop/rspec/expect_output.rb (hunk 5 above)
Comment (backus) [inline lib/rubocop/cop/rspec/expect_output.rb]: Sure I can make that change

File: lib/rubocop/cop/rspec/expect_output.rb (hunk 6 above)
Comment (backus) [inline lib/rubocop/cop/rspec/expect_output.rb]: We only want to flag overwriting of `$stdout` or `$stderr` in the context of an example because we are looking for the pattern of

1. Write to `$stdout`
2. Run task being tested
3. Check value of stdout
4. Reset global variable

If someone is altering either global variable in `before(:all)` or `before(:suite)` then something else is going on that doesn't relate to this cop

File: spec/rubocop/cop/rspec/expect_output_spec.rb (hunk 7 above)
Comment (backus) [inline spec/rubocop/cop/rspec/expect_output_spec.rb]: Explained here https://github.com/backus/rubocop-rspec/pull/294/files#r96149573

File: lib/rubocop/cop/rspec/expect_output.rb (hunk 4 above)
Comment (backus) [inline lib/rubocop/cop/rspec/expect_output.rb]: Will do in a separate PR

Comment (backus) [inline lib/rubocop/cop/rspec/expect_output.rb]: See #296

**Comment by backus** [comment]
@bquorning ready for another pass
